### PR TITLE
[RFC] Generalize ticks in the time HIL, and add support for 24- and 32-bit ticks

### PR DESCRIPTION
**Pull request by gendx**
### Pull Request Overview

This pull request modifies the time HIL to support various width of clock ticks, as an alternative proposal to https://github.com/tock/tock/pull/1413. In particular, support for 24-bit and 32-bit ticks are added, but it could easily be extended to other widths in the future as needed.

- Added a `Ticks` associated type for the `Time` trait, in lieu of the currently unused `W` parameter. This is more consistent with the `Frequency` associated type and is more convenient to work it through dependencies.
- The `Ticks` trait represent a type encapsulating a ticks value (i.e. a 24-bit or 32-bit integer), with support for wrapping arithmetic and conversion to/from `u32`/`usize`.
- Removed the `Time::max_tics(&self)` function. This was unused. Instead, `Ticks::max_value` is now available (and doesn't take any parameter, as it is a constant of the type).
- Added the `Ticks::expired` primitive, to unify an implementation that was duplicated in various capsules.
- Added `Time::ticks_from_ms` and similar utility functions, to avoid the need for dealing with frequencies in all the capsules. This centralizes the arithmetic logic in one place, to reduce code duplication.
- Added a new `Alarm::set_alarm_from_now` function, to support the common `self.set_alarm(self.now().wrapping_add(duration))` pattern and reduce code duplication.

### Testing Strategy

This pull request was tested with:
- `ci-travis`
- new unit tests added in `kernel::hil::timer`.

I haven't done much more tests for now, to first gather feedback on the implementation.

### TODO or Help Wanted

This pull request still needs:
- [ ] Checking that the widths are correct for all clocks.
- [ ] Double-checking that the affected capsules are still correct.

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

File: kernel/src/hil/time.rs [hunk 1]
```diff
@@ -2,20 +2,31 @@
...
     /// Returns the current time in hardware clock units.
-    fn now(&self) -> W;
+    fn now(&self) -> Self::Ticks;
 
-    /// Returns the wrap-around value of the clock.
-    ///
-    /// The maximum value of the clock, at which `now` will wrap around. I.e., this should return
-    /// `core::u32::MAX` on a 32-bit-clock, or `(1 << 24) - 1` for a 24-bit clock.
-    fn max_tics(&self) -> W;
+    fn ticks_from_seconds(s: u32) -> Self::Ticks {
```
Comment (phil-levis) [inline kernel/src/hil/time.rs]: I'm a little wary of these conversion routines, particularly because ticks_from_ms(ticks_to_ms(x)) won't necessarily compute the identity, and because there are floating point operations. 

It also can lead to runtime failures, because not every tick frequency can support every unit. For example, if the Alarm is based off a 32kHz clock and code requests a 10us Alarm. I realize this problem existed with the previous trait as well. Is there any way to parameterize the Alarm trait by the tick unit, such that, for example, the UDP test has a ms alarm but I2C has a us alarm? A ms alarm can just be a scaling of the us one.

Comment (gendx) [reply kernel/src/hil/time.rs]: > I'm a little wary of these conversion routines, particularly because ticks_from_ms(ticks_to_ms(x)) won't necessarily compute the identity, and because there are floating point operations.

1. `ticks_from_ms(ticks_to_ms(x))` will definitely pretty much never compute the identity, whatever the formula is (using floating point or integer arithmetic), unless the frequency is a "nice" number (like 1, or powers of two for "reasonable" values of x that don't trigger an overflow), or unless `ticks_to_ms(x)`'s result stores the original `x`. But is it unexpected? If one wants to keep the precision, the operations must be done in ticks.
Actually, the only use of `ticks_to_ms` I found was in `capsules/src/ieee802154/xmac.rs`, and it's questionable whether this conversion is needed there, but I don't have knowledge of this protocol & driver to assess that.
The other uses of `ticks_from_ms` all sound reasonable to me (most seem like heuristics, or "average" values from a data sheet). It's unclear though whether the value should be rounded up instead of down, so that the waiting time is indeed at least N ms, instead of N-1 < x <= N.
1. There was a lot of code duplication so I took the approach of consistently using floating point arithmetic to avoid overflows on multiplication. Given that at least one driver was using floating point arithmetic I assumed that it was fine, but we could also use integer arithmetic instead, or provide two variants of the function.
The caveat of integer arithmetic is that "ms * frequency" can overflow before dividing by 10^3. On the other hand, one can divide by 10^3 before multiplying by the frequency, but then the precision is rounded to full seconds. A more accurate formula is `(ms / 1000) * f + ((ms % 1000) * f) / 1000` - assuming that `1000 * f` doesn't overflow (if you convert from microseconds instead of milliseconds it will overflow).
Another alternative is to use 64-bit arithmetic. It can be evaluated even on 32-bit CPUs (at the expense of a few more instructions), and given that the divisor is a compile-time constant, the compiler should be able to turn the division into a combination of multiplication(s), shift(s) and other simple arithmetic operations (example, on x86_64: https://godbolt.org/z/7nvfWw). Also, with a compile-time constant number of milliseconds (which is the most common case), the compiler can just evaluate that and bake a constant in the binary. So 64-bit arithmetic would be the best alternative to floating point IMO

What do you think?

> It also can lead to runtime failures, because not every tick frequency can support every unit. For example, if the Alarm is based off a 32kHz clock and code requests a 10us Alarm. I realize this problem existed with the previous trait as well. Is there any way to parameterize the Alarm trait by the tick unit, such that, for example, the UDP test has a ms alarm but I2C has a us alarm? A ms alarm can just be a scaling of the us one.

That would be possible, although out of the scope of this pull-request. The core goal of this pull-request is fixing 24-bit ticks, so that the alarms based on nRF52's RTC work as intended, and it's already a large enough change to fix an orthogonal issue.

**Comment by bradjc** [comment]
I am excited about the new helper functions (seems better than just blindly copying the math).

But I'm not sure about the need to change the include paths in so many files.

File: kernel/src/hil/time.rs (hunk 1 above)
Comment (phil-levis-google) [reply kernel/src/hil/time.rs]: >    Actually, the only use of `ticks_to_ms` I found was in `capsules/src/ieee802154/xmac.rs`, and it's questionable whether this conversion is needed there, but I don't have knowledge of this protocol & driver to assess that.

Yeah, this is exactly the kind of use I was worried about. The protocol is trying to figure out when to turn the radio on. The more accurately it can do this, the less energy it will use. What the code does is take the ticks remaining, turn them into ms, then convert back from ms to ticks. I think it's pretty simple to fix this (change `set_timer_ms` to `set_timer_ticks` and just use ticks), and in that case there's no need for `ticks_to_ms`. 

>    The other uses of `ticks_from_ms` all sound reasonable to me (most seem like heuristics, or "average" values from a data sheet). It's unclear though whether the value should be rounded up instead of down, so that the waiting time is indeed at least N ms, instead of N-1 < x <= N.

Round up. Or just round.

> 2. There was a lot of code duplication so I took the approach of consistently using floating point arithmetic to avoid overflows on multiplication. Given that at least one driver was using floating point arithmetic I assumed that it was fine, but we could also use integer arithmetic instead, or provide two variants of the function.
>    The caveat of integer arithmetic is that "ms * frequency" can overflow before dividing by 10^3. On the other hand, one can divide by 10^3 before multiplying by the frequency, but then the precision is rounded to full seconds. A more accurate formula is `(ms / 1000) * f + ((ms % 1000) * f) / 1000` - assuming that `1000 * f` doesn't overflow (if you convert from microseconds instead of milliseconds it will overflow).
>    Another alternative is to use 64-bit arithmetic. It can be evaluated even on 32-bit CPUs (at the expense of a few more instructions), and given that the divisor is a compile-time constant, the compiler should be able to turn the division into a combination of multiplication(s), shift(s) and other simple arithmetic operations (example, on x86_64: https://godbolt.org/z/7nvfWw). Also, with a compile-time constant number of milliseconds (which is the most common case), the compiler can just evaluate that and bake a constant in the binary. So 64-bit arithmetic would be the best alternative to floating point IMO
> 
> What do you think?

64 bit is the way to go. I think we encountered something like this before in timer code (I'm struggling to find where) and the conclusion was to use 64-bit arithmetic rather than do tricks with 32-bit arithmetic.

> > It also can lead to runtime failures, because not every tick frequency can support every unit. For example, if the Alarm is based off a 32kHz clock and code requests a 10us Alarm. I realize this problem existed with the previous trait as well. Is there any way to parameterize the Alarm trait by the tick unit, such that, for example, the UDP test has a ms alarm but I2C has a us alarm? A ms alarm can just be a scaling of the us one.
> 
> That would be possible, although out of the scope of this pull-request. The core goal of this pull-request is fixing 24-bit ticks, so that the alarms based on nRF52's RTC work as intended, and it's already a large enough change to fix an orthogonal issue.

I agree.

Comment (gendx) [reply kernel/src/hil/time.rs]: Migrated to `u64` arithmetic, and removed `ticks_to_ms`. Please confirm that the `ieee802154/xmac` capsule is still correct.

**Comment by gendx** [comment]
> But I'm not sure about the need to change the include paths in so many files.

In many cases there were inconsistencies with some paths fully included (`kernel::hil::time::Frequency`), whereas others where pulled from `time::`, `hil::time::` or with the full path from the `kernel::` (even within a file). Given that I had to remove some now-unused imports (typically `Frequency`), I figured it would be clearer to have a more consistent import pattern.

But I can simplify some of that if that helps for reviewing this.

**Comment by phil-levis** [comment]
@alevy Can you take a look at this?

File: kernel/src/hil/time.rs (hunk 1 above)
Comment (niklasad1) [reply kernel/src/hil/time.rs]: hrm, could we use `core::time::Duration` instead of `ticks_from_xx` instead?

**Comment by phil-levis** [comment]
bors try

**Comment by phil-levis** [comment]
@gendx With ibex/OT now in this doesn't build. Would you like to bring them in sync or shall I? Then I think this is good to go.

**Comment by pfmooney** [comment]
> @gendx With ibex/OT now in this doesn't build. Would you like to bring them in sync or shall I? Then I think this is good to go.

With this PR rebased on master, I've fixed the build errors for Ibex (and the Alarm test capsule) in oxidecomputer/tock@d1a8a38e8c1f66b8344741813d333893bd57d291 and oxidecomputer/tock@84cae359c300806dae85f6b0c78af7ea30808321.

FYI: While they fix the obvious build errors, I haven't had an opportunity to test on hardware yet.

File: kernel/src/hil/time.rs [hunk 2]
```diff
@@ -86,10 +93,26 @@ pub trait Alarm<'a, W = u32>: Time<W> {
...
+    /// [`Client#fired`](trait.Client.html#tymethod.fired) is signaled when the
+    /// alarm is elapsed.
+    ///
+    /// # Examples
+    ///
+    /// ```ignore
+    /// let delta = 1337;
+    /// alarm.set_alarm_from_now(delta);
+    /// ```
+    fn set_alarm_from_now(&self, duration: Self::Ticks) {
```
Comment (alevy) [inline kernel/src/hil/time.rs]: I actually don't think we want to encourage clients to use this kind of helper method. In particular, it's very important for time-related applications/drivers to reason precisely about _when_ `now()` was taken.

I do understand the desire to hide this relatively gross but common line of code.

Perhaps an alternative would be to have `Ticks` implement `core::ops::Add`, `core::ops::Sub` using wrapping logic such that clients can just use `+`, `-`, etc? In the same way `core::num::Wrapping<T>` does.

File: kernel/src/hil/time.rs [hunk 3]
```diff
@@ -175,3 +198,254 @@ pub trait TimerClient {
     /// Callback signaled when the timer's clock reaches the specified interval.
     fn fired(&self);
 }
+
+pub trait Ticks: Clone + Copy + From<u32> {
+    fn into_usize(self) -> usize;
```
Comment (alevy) [inline kernel/src/hil/time.rs]: Why not just require an implementation of the `Into` and/or `From` traits?

File: kernel/src/hil/time.rs [hunk 4]
```diff
@@ -175,3 +198,254 @@ pub trait TimerClient {
...
+
+    /// Returns the wrap-around value of the clock.
+    ///
+    /// The maximum value of the clock, at which `now` will wrap around. I.e., this should return
+    /// `core::u32::MAX` on a 32-bit-clock, or `(1 << 24) - 1` for a 24-bit clock.
+    fn max_value() -> Self;
+}
+
+#[derive(Clone, Copy)]
+pub struct Ticks32Bits(u32);
```
Comment (alevy) [inline kernel/src/hil/time.rs]: Seems we could just implement `Ticks` for `u32` directly, no?

File: kernel/src/hil/time.rs [hunk 5]
```diff
@@ -175,3 +198,254 @@ pub trait TimerClient {
...
+        now.wrapping_sub(reference).0 >= when.wrapping_sub(reference).0
+    }
+
+    fn max_value() -> Self {
+        Self(Self::MAX_VALUE)
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct Ticks24Bits(u32);
```
Comment (alevy) [inline kernel/src/hil/time.rs]: Perhaps this should be called something like `U24` instead?

File: kernel/src/hil/time.rs [hunk 6]
```diff
@@ -175,3 +198,254 @@ pub trait TimerClient {
...
+
+pub trait Ticks: Clone + Copy + From<u32> {
+    fn into_usize(self) -> usize;
+    fn into_u32(self) -> u32;
+
+    fn wrapping_add(self, other: Self) -> Self;
+    fn wrapping_sub(self, other: Self) -> Self;
+
+    /// Check whether `now` is after or equal to `when`, w.r.t. a `reference` time point.
+    fn expired(reference: Self, now: Self, when: Self) -> bool;
```
Comment (alevy) [inline kernel/src/hil/time.rs]: Isn't the implementation for this always the same? Should this be a helper instead of a method? also, it might be more clear what each argument means if it is named slightly differently:

```rust
fn gt_or_eq<T: Ticks>(reference: T, lhs: T, rhs: T) -> bool
```

File: kernel/src/hil/time.rs [hunk 7]
```diff
@@ -175,3 +198,254 @@ pub trait TimerClient {
...
+        assert_eq!(Time32Bits16MHz::ticks_from_us(1).into_u32(), 16);
+        assert_eq!(Time32Bits16MHz::ticks_from_us(10).into_u32(), 160);
+        assert_eq!(Time32Bits16MHz::ticks_from_us(100).into_u32(), 1_600);
+        assert_eq!(Time32Bits32KHz::ticks_from_us(1).into_u32(), 0);
+        assert_eq!(Time32Bits32KHz::ticks_from_us(10).into_u32(), 0);
+        assert_eq!(Time32Bits32KHz::ticks_from_us(100).into_u32(), 3);
+    }
+
+    #[test]
+    fn test_expired_reference_zero() {
```
Comment (alevy) [inline kernel/src/hil/time.rs]: It would be good to also test non-zero references

File: kernel/src/hil/time.rs (hunk 1 above)
Comment (alevy) [reply kernel/src/hil/time.rs]: `u64` is definitely better than floating point (the code bloat alone from software floating point would be a huge problem a big problem on non-M4F platforms).

I agree with @niklasad1 that if there isn't a reason _not_ to use `Duration` it probably makes sense to unify all of these with `ticks_from_duration`.

Finally, this is also a candidate for being a helper as opposed to a method.

File: kernel/src/hil/time.rs (hunk 2 above)
Comment (bradjc) [reply kernel/src/hil/time.rs]: This actually changes the HIL, rather than just supports different counter widths.

File: kernel/src/hil/time.rs (hunk 6 above)
Comment (bradjc) [reply kernel/src/hil/time.rs]: I agree that it's not clear what it would mean for a number of bits (or counter width) to be expired.

File: kernel/src/hil/time.rs [hunk 8]
```diff
@@ -2,20 +2,27 @@
...
-    /// Returns the wrap-around value of the clock.
-    ///
-    /// The maximum value of the clock, at which `now` will wrap around. I.e., this should return
-    /// `core::u32::MAX` on a 32-bit-clock, or `(1 << 24) - 1` for a 24-bit clock.
-    fn max_tics(&self) -> W;
+    fn ticks_from_seconds(s: u32) -> Self::Ticks {
+        Self::Ticks::from(s * Self::Frequency::frequency())
+    }
+
+    fn ticks_from_ms(ms: u32) -> Self::Ticks {
```
Comment (bradjc) [inline kernel/src/hil/time.rs]: Should be `ticks_from_milliseconds` or the function above should change.

**Comment by pfmooney** [comment]
> Is this PR holding up any other work? I think given this is a second revision of a HIL in a relatively short time (although we agreed that we hadn't necessarily figured out the width issue last time) we can take some time on the updates.
> 
> I'm most concerned about the number of moving parts in the HIL with limited comments describing why the complexity is necessary. I'm happy to take a first pass at describing what the different aspects are for.

I agree that it's a good opportunity to scrutinize the Time/Alarm HIL.  Looking closer at the RISC-V MachineTimer, one will find that it operates differently than practically all of the other uC timer peripherals: It matches/fires on `time >= compare` rather than `time == compare`.  Assuming first-class support for RISC-V is a goal of Tock, it would be nice for the Alarm model to be able to handle such cases.

It wouldn't be shocking if such changes reached all the way to the syscall interface, making the effort a good fit to dovetail into "Tock 2.0" as [mentioned recently on tock-dev](https://groups.google.com/d/msg/tock-dev/jicoPcapHNU/FM_ribgHAgAJ).

**Comment by bradjc** [comment]
I added a first-pass on documenting more of the HIL. I think refactoring the ticks types makes sense, and then I can document those as well.

**Comment by phil-levis** [comment]
> I agree that it's a good opportunity to scrutinize the Time/Alarm HIL. Looking closer at the RISC-V MachineTimer, one will find that it operates differently than practically all of the other uC timer peripherals: It matches/fires on `time >= compare` rather than `time == compare`. Assuming first-class support for RISC-V is a goal of Tock, it would be nice for the Alarm model to be able to handle such cases.

I've been thinking about this for the past few days, and I'm hard pressed to come up with a good way to generalize >= and ==. I can understand the appeal of having a counter that will never wrap around, but having >= means doing nothing will result in repeated interrupts. I'm leaning towards the idea that the RISC-V implementation of Alarm should just disable the interrupt until the Alarm is reset. Do you see a way to generalize it? If so we should discuss on -dev.

**Comment by pfmooney** [comment]
> I've been thinking about this for the past few days, and I'm hard pressed to come up with a good way to generalize >= and ==.

Currently the `Alarm` HIL (and to a lesser extent `Time`) is rather leaky as an abstraction, making little effort to hide the details of its underlying hardware resource from consumers.  For use cases where only a minimal shim over the hardware is desired, this would be fine.  This seems at odds, however, with the flexibility of the Capsule system in Tock.  Having timekeeping facilities which do not betray the details of their underlying implementation seems like a valuable resource for such nominally hardware-agnostic code.

> Do you see a way to generalize it? If so we should discuss on -dev.

I certainly have enough to draw up a straw man for discussion.  I'll work on a write-up for tock-dev.

**Comment by pfmooney** [comment]
> I certainly have enough to draw up a straw man for discussion. I'll work on a write-up for tock-dev.

The write-up has been [posted to tock-dev](https://groups.google.com/d/msg/tock-dev/CWecIZq7jqg/AvLtMBedEwAJ).

**Comment by phil-levis** [comment]
One question here -- this is clearly a step in the right direction. Do we want to hold off on the full Alarm/Time redesign, or merge this now?

**Comment by bradjc** [comment]
I think merging a smaller version of this that only improves the support for timer widths (and doesn't change the core api) is fine.

**Comment by gendx** [comment]
> I think merging a smaller version of this that only improves the support for timer widths (and doesn't change the core api) is fine.

Could you be more specific about what to include in a "smaller" version?
Changes boil down to the following:
- Adding the `Ticks` trait to model various widths.
- Replacing the `W` parameters in the time HIL by the `Ticks` trait. The previous `W` wasn't constrained, and really worked only for `W = u32`. So it's a change in the HIL, but given that the previous HIL was broken in that it only supported `u32`, I don't see why we shouldn't update the HIL.
- Cleaning up imports in capsules. Sure that's quite a bit of diff lines, but on the other hand makes things more uniform across all capsules.
- Removing `max_ticks()` from `kernel::hil::Time`. This wasn't used anywhere, and I think that the `Ticks` trait models that better.
- Adding `ticks_from_ms` and `set_alarm_from_now` to better model existing patterns in capsules. That's indeed a change to the core API, but I think in the good direction as it avoids plenty of code duplication and reduces the surface exposed by the API (no need to expose low-level details such as the frequency or when is `now`) - given all the current usages. One can imagine other patterns to use timers, but these new functions are all that's needed to reproduce current behavior - and more functions can be added to the API later if needed.

**Comment by bradjc** [comment]
I would like to merge this with all of Amit's comments addressed and the changes to the HIL restricted only to the width issue. Changes to the overall time design are still in discussion, and we tend to avoid small API changes when we know a larger design is in the works.

**Comment by gendx** [comment]
> I would like to merge this with all of Amit's comments addressed

I've noted these comments, but in the interest of saving time, I don't want to invest time on addressing them until we've reached an agreement on the other points.

> and the changes to the HIL restricted only to the width issue. Changes to the overall time design are still in discussion, and we tend to avoid small API changes when we know a larger design is in the works.

The proposed HIL changes are:
1. consistently supporting the new `Ticks` trait (apart from naming comments, I don't see it controversial to do them here),
1. consistently converting ticks to/from seconds (of divisions of a second),
1. the new `set_alarm_from_now` interface.

Although to/from seconds conversions are not strictly necessary here, I see them as useful - more readable for users of this API, a single conversion formula so easier to get correct. Please let me know if that's controversial, but it would be IMO detrimental to push that to another pull-request.

As far as I understand, the most controversial is the new `set_alarm_from_now` API. However, in the light of #1651, the current HIL is problematic (so I don't see why we should limit ourselves in changing the time HIL just because it was changed recently). The "read now & set now+delay" pattern isn't sound, because as explained in #1651, by the time we get down to the chip, the "now+delay" timestamp may already be in the past.

Given that all capsules are effectively using alarms in a `set_alarm_from_now` way, I think it makes sense to formalize it for now. By simplifying the public API of the time HIL, this will allow more flexibility under the hood to implement things correctly and fix #1651 (without having to refactor all the capsules when prototyping each change on the chips/kernel side of things). If we want to discourage clients from using the `set_alarm_from_now` pattern, we should later remove this function and provide a suitable upgrade path, but the current behavior is consistently a `set_alarm_from_now` pattern.

We can always update the HIL again once we have a clear way of fixing #1651, as well as when we have a clear way of supporting the 64-bit timers on other architectures, and when we have a clear way of fixing the `set_alarm_from_now` pattern. But given that the time HIL is still in a work-in-progress state (given all the bugs related to the current implementation), I don't see it problematic to make multiple changes on this HIL in a short period of time on the development branch of Tock. I think that the HIL should stabilize on the next release of Tock instead of after each pull-request.

Given that the changes required to support ticks require some refactoring in the capsules anyway, I don't see it useful not to also introduce the `set_alarm_from_now` API, in order to reduce friction in refactoring that again.

Feel free to propose a smaller version of this pull-request if you disagree. Or a way to fix #1651 first if you think it's higher priority.

**Comment by bradjc** [comment]
I'm looking for help from others on this pull request. The time/alarm discussion is outside of my expertise and I don't have the bandwidth to focus on it and its intricacies. But this PR has been lingering for months and I want to see it progress (and I know that we have agreed that better width handling is important, so that is an easy merge). But I am not the person to work out the details. Sorry @gendx for being annoying in this thread and my lack of useful comments.

**Comment by bradjc** [comment]
Closing without judgement, just expecting the alarm-related PRs to be handled with a new PR from the "time taskforce" (I don't know if that is the group's actual name).
